### PR TITLE
Null safety migration

### PR DESCRIPTION
**Pull request by justacid**
Hi there!

I updated stomp_dart_client to support null-safety. Since this is a breaking change anyway and a version bump is recommended, I took the liberty to improve type-safety by removing any incidental dynamic's in the callback function signatures (using typedefs for clarity) and in tests. I also changed the 'int delay params' in the `StompConfig` to `Duration`s and formatted the code. 

All tests are passing and it works within my project. Let me know what you think! 
Cheers :)

### CODE REVIEW

**Comment by KammererTob** [comment]
Hey! Thanks for the PR! I will have a look next week.

File: lib/stomp.dart [hunk 1]
```diff
@@ -49,60 +41,77 @@ class StompClient {
...
-      @required Function(StompFrame) callback,
-      Map<String, String> headers}) {
-    return _handler.subscribe(
-        destination: destination, callback: callback, headers: headers);
+  StompUnsubscribe subscribe({
+    required String destination,
+    required StompFrameCallback callback,
+    Map<String, String>? headers,
+  }) {
+    if (_handler != null) {
```
Comment (KammererTob) [inline lib/stomp.dart:69]: I think we should use `connected` here and also throw an Exception in the else branch. This way the caller knows that a subscription or send message was never really transmitted.

File: lib/stomp_parser.dart [hunk 2]
```diff
@@ -213,10 +229,10 @@ class StompParser implements Parser {
         .replaceAll(RegExp(r'\r'), '\\r');
   }
 
-  Map<String, String> _escapeHeaders(Map<String, String> headers) {
+  Map<String, String> _escapeHeaders(Map<String?, String> headers) {
```
Comment (KammererTob) [inline lib/stomp_parser.dart:232]: I might be blind, but why is this signature needed? Isn't this only ever called with headers filled from the StompFrame and the field inside StompFrame has a non-null signature.

**Review by KammererTob — COMMENTED** [review]
Thanks for this great PR. I've added some comments and open for discussion about anything i noted.

File: lib/stomp.dart (hunk 1 above)
Comment (justacid) [reply lib/stomp.dart:69]: Something like this?

```dart
if (connected) {
  return _handler!.subscribe(
    destination: destination,
    callback: callback,
    headers: headers,
  );
}

throw StompClientDisconnectedException();
```

I'm not entirely happy with this, because the `connected` getter hides the null-check from the reader, which makes it not obvious that `_handler!` is correct. And unforunately, because Dart can not promote nullable instance fields automatically to a non-nullable type, we can not get rid of the '!' here.

An alternative could be

```dart
final handler = _handler;
if (handler == null || !handler.connected) {
  throw StompClientDisconnectedException();
}

return handler.subscribe(
  destination: destination,
  callback: callback,
  headers: headers,
);
```

which feels icky to me too. What do you think?

File: lib/stomp_parser.dart (hunk 2 above)
Comment (justacid) [reply lib/stomp_parser.dart:232]: You are right, this does not need to be nullable - just an oversight!

File: lib/stomp.dart [hunk 3]
```diff
@@ -49,60 +41,77 @@ class StompClient {
...
-    }, onWebSocketDone: () {
-      config.onWebSocketDone();
-
-      if (_isActive) {
-        _scheduleReconnect();
-      }
-    }));
-    _handler.start();
+      config: config.copyWith(
+        onConnect: (_, frame) {
```
[the remark this answers is not in the document]
Comment (justacid) [reply lib/stomp.dart:45]: Not quite sure, what you propose here. Do you mean passing the separate (internal) callback to the `StompHandler` directly, like so?

```dart
StompHandler({required this.config, required this.onConnected}) { ... }
```

Another option would be to simply remove the client parameter alltogether. It seems to be that everybody needs to keep a reference to the `StompClient` around anyway. Or is there a use case I'm not seeing?

File: lib/parser.dart [hunk 4]
```diff
@@ -1,7 +1,7 @@
 import 'package:stomp_dart_client/stomp_frame.dart';
 
 abstract class Parser {
-  bool escapeHeaders;
+  late bool escapeHeaders;
```
[the remark this answers is not in the document]
Comment (justacid) [reply lib/parser.dart:4]: Right, the sock_js_parser is not using it, but the flag is dynamically injected in the `_onConnectFrame` handler, after one of both parsers is already instantiated:

```dart
if (frame.headers['version'] != '1.0') {
  _parser.escapeHeaders = true;
} else {
  _parser.escapeHeaders = false;
}
```

Therefore, the only easy way around this I see right now would be an explicit `is StompParser` check, but this is probably brittle too. I'd tend to leave it as is. Another option would be to drop support for the 1.0 protocol, since that is the only reason this is even needed (I think?), but I guess we would not want that.

File: lib/stomp.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (KammererTob) [reply lib/stomp.dart:45]: Yeah i think you are right. We can just remove it. I don't use it in my project aswell and i did not come up with any use case where this would be helpful.

File: lib/parser.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (KammererTob) [reply lib/parser.dart:4]: Okay. Did not see this. We leave it then.

File: lib/stomp.dart (hunk 1 above)
Comment (KammererTob) [reply lib/stomp.dart:69]: Couldn't we maybe just  do `handler!.` since we assume that any other state than non-null should throw an exception either way? This way a null exception is thrown when the handler is null and some sort of WebSocketException (i am not sure which one) would be thrown if the underlying channel/sink is no longer open?

Comment (justacid) [reply lib/stomp.dart:69]: Honestly, I'm not sure of the underlying exceptions... I observed in my crashlytics that sometimes `_transmit` fails, because `channel` was null in the `StompHandler`: `NoSuchMethodError: The getter 'sink' was called on null. Receiver: null Tried calling: sink`, but I didn't understand why this would happen exactly (and it didn't really matter to me). This happend if I called `subscribe` in the `onConnected` callback - though it did not reappear after explicitly checking the client for both null and an active connection before calling `subscribe`. I never managed to reproduce this locally.

In the current PR there is a null-check around the `channel` in `_transmit`, which prevents this error (and silently ignores it) - I guess if you remove the null-check it would simply crash with a 'null check operator used on a null value' error instead (in this case). Though, as mentioned, I never could actually make it crash locally, hence I'm pretty much guessing here.

Otherwise, I would still prefer for subscribe/send and ack/nack to either 'ignore the callee', if not connected or otherwise throw an explicit exception that tells me whats wrong (i.e. a StompClientDisconnectedException).

BTW, I just realized that `channel` in `StompHandler` is public, can we make this private?

File: lib/stomp.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (justacid) [reply lib/stomp.dart]: Allright, I removed it!

File: lib/stomp.dart (hunk 1 above)
Comment (KammererTob) [reply lib/stomp.dart:69]: I am just guessing here, but the only way i see channel being null when calling subscribe inside the `onConnected` callback is if the `StompHandler` was replaced in the `StompClient ` after calling `start` on it (maybe calling `activate` twice?). This could maybe be fixed if we dispose the handler before opening a new one (although i am not sure of any ramifcations this might have). But we are getting a bit OT.

I think we should make sure that the user is informed when a call to subscribe/send/ack/nack is made with an inactive/disconnected client and not silently ignore it. For me this means:
- When any of those methods is called before the client was initialized (so before `activate` was called): BadState
- When any of those methods is called and the client is no longer connected: Disconnected

What i am unsure of is if we should supply our own Exceptions or just let dart/WebSocketChannel handle this for us like it was in the code before (NoSuchMethodError and whatever Exception WebSocketChannel throws when we try add something to the sink while it is closed. I would assume a BadState)

Comment (KammererTob) [reply lib/stomp.dart:69]: Your arguments make sense. I guess we probably then need two exceptions:
1. BadStateException: Calling send/subscribe/ack/nack while `handler` is null. The only reason for the handler being null at that point is that activate wasn't called or if the beforeConnect future is still running.
2. StompClientDisconnectedException (or something along those lines): Calling send/subscribe/ack/nack while being disconnected. This should probably be thrown inside the `_transmit` method.

Comment (KammererTob) [reply lib/stomp.dart:69]: Perfect. Thanks again for your effort!

Comment (justacid) [reply lib/stomp.dart]: So, I justed looked into it, and for now I went with a single `StompBadStateException`:

* In `_transmit` I can not check if the client is connected because `_transmit` itself is used to establish the connection between the client and the server. Of course, I can check whether the websocket channel is null, but this is really, as you mentioned before, a 'bad state' that should not actually happen.
* I got rid of the idea of a separate `DisconnectedException` after trying it out, because all exceptions are ultimately about the connection state, hence having an extra exception type does not seem semantically meaningful. Theres nothing different to act on. The only thing it does is force the callee to catch two separate exceptions instead of one.
* Basically, I wrapped `channel!.sink.add` in a try-catch and convert all exceptions into a `StompBadStateException`. In the `StompClient` I throw the exception for when the `_handler` is null.
* I also tried to add meaningful error messages and added two test cases for the handler and the client.

Please have a look and tell me what you think. :)

PS: I also made `channel` private in the `StompHandler`.

Comment (KammererTob) [reply lib/stomp.dart]: Very nice. Thank you! I will merge this PR and release a new version this weekend or early next week.

Comment (justacid) [reply lib/stomp.dart]: Thank you too! 👍🏻
